File: Raven.Abstractions/Commands/PatchCommandData.cs
//-----------------------------------------------------------------------
// <copyright file="PatchCommandData.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System.Linq;
using Raven.Abstractions.Data;
using Raven.Json.Linq;

namespace Raven.Abstractions.Commands
{
	///<summary>
	/// A single batch operation for a document PATCH
	///</summary>
	public class PatchCommandData : ICommandData
	{
		/// <summary>
		/// Gets or sets the patches applied to this document
		/// </summary>
		/// <value>The patches.</value>
		public PatchRequest[] Patches{ get; set;}

		/// <summary>
		/// Gets or sets the patches to apply to a default document if the document is missing
		/// </summary>
		/// <value>The patches.</value>
		public PatchRequest[] PatchesIfMissing { get; set; }

		/// <summary>
		/// Gets the key.
		/// </summary>
		/// <value>The key.</value>
		public string Key
		{
			get; set;
		}

		/// <summary>
		/// Gets the method.
		/// </summary>
		/// <value>The method.</value>
		public string Method
		{
			get { return "PATCH"; }
		}

		/// <summary>
		/// Gets or sets the etag.
		/// </summary>
		/// <value>The etag.</value>
		public Etag Etag
		{
			get; set;
		}

		/// <summary>
		/// Gets the transaction information.
		/// </summary>
		/// <value>The transaction information.</value>
		public TransactionInformation TransactionInformation
		{
			get; set;
		}

		/// <summary>
		/// Gets or sets the metadata.
		/// </summary>
		/// <value>The metadata.</value>
		public RavenJObject Metadata
		{
			get; set;
		}

		public RavenJObject AdditionalData { get; set; }

		/// <summary>
		/// Translate this instance to a Json object.
		/// </summary>
		public RavenJObject ToJson()
		{
			var ret = new RavenJObject
			       	{
						{"Key", Key},
						{"Method", Method},
						{"Patches", new RavenJArray(Patches.Select(x => x.ToJson()))},
						{"Metadata", Metadata},
						{"AdditionalData", AdditionalData}
					};
			if (Etag != null)
				ret.Add("Etag", Etag.ToString());
			if (PatchesIfMissing != null && PatchesIfMissing.Length > 0)
				ret.Add("PatchesIfMissing", new RavenJArray(PatchesIfMissing.Select(x => x.ToJson())));
			return ret;
		}
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/abstractions/commands/PatchCommandData.java
package raven.abstractions.commands;

import raven.abstractions.basic.SharpEnum;
import raven.abstractions.data.Etag;
import raven.abstractions.data.HttpMethods;
import raven.abstractions.data.PatchRequest;
import raven.abstractions.data.TransactionInformation;
import raven.abstractions.json.linq.RavenJArray;
import raven.abstractions.json.linq.RavenJObject;
import raven.abstractions.json.linq.RavenJValue;

public class PatchCommandData implements ICommandData {

  private PatchRequest[] patches;
  private PatchRequest[] patchesIfMissing;
  private String key;
  private Etag etag;
  private TransactionInformation transactionInformation;
  private RavenJObject metadata;
  private RavenJObject additionalData;

  public RavenJObject getAdditionalData() {
    return additionalData;
  }

  public Etag getEtag() {
    return etag;
  }

  public String getKey() {
    return key;
  }

  public RavenJObject getMetadata() {
    return metadata;
  }

  public HttpMethods getMethod() {
    return HttpMethods.PATCH;
  }

  public PatchRequest[] getPatches() {
    return patches;
  }

  public PatchRequest[] getPatchesIfMissing() {
    return patchesIfMissing;
  }

  public TransactionInformation getTransactionInformation() {
    return transactionInformation;
  }

  public void setAdditionalData(RavenJObject additionalData) {
    this.additionalData = additionalData;
  }

  public void setEtag(Etag etag) {
    this.etag = etag;
  }

  public void setKey(String key) {
    this.key = key;
  }

  public void setMetadata(RavenJObject metadata) {
    this.metadata = metadata;
  }

  public void setPatches(PatchRequest[] patches) {
    this.patches = patches;
  }

  public void setPatchesIfMissing(PatchRequest[] patchesIfMissing) {
    this.patchesIfMissing = patchesIfMissing;
  }

  public void setTransactionInformation(TransactionInformation transactionInformation) {
    this.transactionInformation = transactionInformation;
  }

  @Override
  public RavenJObject toJson() {
    RavenJObject ret = new RavenJObject();
    ret.add("Key", new RavenJValue(key));
    ret.add("Method", new RavenJValue(SharpEnum.value(getMethod())));

    RavenJArray patchesArray = new RavenJArray();
    for (PatchRequest patchRequest: patches) {
      patchesArray.add(patchRequest.toJson());
    }
    ret.add("Patches", patchesArray);
    ret.add("AdditionalData", additionalData);
    if (etag != null) {
      ret.add("Etag", new RavenJValue(etag));
    }
    if (patchesIfMissing != null && patchesIfMissing.length > 0) {
      RavenJArray patchesIfMissingArray = new RavenJArray();
      for (PatchRequest patchRequest: patchesIfMissing) {
        patchesIfMissingArray.add(patchRequest.toJson());
      }
      ret.add("PatchesIfMissing", patchesIfMissingArray);
    }
    return ret;
  }

}
